### PR TITLE
Diagnostic counters

### PR DESCRIPTION
**Pull request by toddmowen**
~~Just creating this PR to share with @SamRoberts.~~

Ready for review now.

This PR provides a new method `withCounter(name: String)` on any TypedPipe (available after importing the implicit conversion `CoppersmithStats.fromTypedPipe`). Code in the scalding subproject should call this whenever the number of records could change, e.g. after filter, join, etc. See the TROUBLESHOOTING file for an example of what the log output looks like.

The following workarounds should possibly be pushed upstream:
- [ ] slf4j-test: loggers are not serializable
- [ ] uniform: putting hive-exec last on the classpath helps avoid surprises

And of course there is also this opportunity...
- [ ] scalding: if bugs around counters are fixed, then several workarounds in this PR can be removed

### CODE REVIEW

File: scalding/src/main/scala/commbank/coppersmith/scalding/CoppersmithStats.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      _                  <- Execution.from(logCounters(counters))
+    } yield result.get  // any exception caught by the above recoverWith is rethrown here
+  }
+
+  /** Log (at INFO level) all coppersmith counters found in the passed [[com.twitter.scalding.ExecutionCounters]]. */
+  def logCounters(counters: ExecutionCounters): Unit = {
+    if (counters.keys.isEmpty) {
+      log.warn("Hadoop counters not available (usually caused by job failure)")
+    }
+    else {
```
Comment (dkristian) [inline scalding/src/main/scala/commbank/coppersmith/scalding/CoppersmithStats.scala]: Minor formatting: combine above two lines

**Comment by dkristian** [comment]
Ok, I've merged this with the latest master, and got some counters being logged on DS. The load counts are exactly as expected, but the write count is 25% short (806811 actual, 615750 reported). This seems a little high to be within the apparent known shortcomings with scaldings counters to me. Happy to go through my methodology when you have time.

Before giving a thumbs up on the code, I'd be keen to see it merged with the latest master branch, as there have been some changes in the same areas. My attempt at rebasing master in order to test is at https://github.com/CommBank/coppersmith/compare/counters-latest-kd?expand=1

**Comment by toddmowen** [comment]
Merged with master. @dkristian's first merge attempt broke HiveTextSink because (surprisingly) the following snippet causes counters to be lost:

```scala
written     <- CoppersmithStats.logCountersAfter(
                 partitioned.writeThrough(sink).withSubConfig(createUniqueFilenames(_)))
```

while this does not:

```scala
written     <- CoppersmithStats.logCountersAfter(partitioned.writeThrough(sink))
                 .withSubConfig(createUniqueFilenames(_))
```

File: scalding/src/main/scala/commbank/coppersmith/scalding/CoppersmithStats.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        log.info(f"    ${name}%-30s ${value}%10d")
+      }
+    }
+  }
+
+  def fromCounters(counters: ExecutionCounters): List[(String, Long)] =
+    counters.keys.filter(_.group == group).map { key =>
+      val Array(id, name) = key.counter.split(raw"\.", 2)
+      (id.toLong, name, key)
+    }.toList.sortBy(_._1).map { case (id, name, key) => (name, counters(key)) }
```
Comment (dkristian) [inline scalding/src/main/scala/commbank/coppersmith/scalding/CoppersmithStats.scala]: Very minor: you could drop the `id` binding here:
```scala
    }.toList.sortBy(_._1).map { case (_, name, key) => (name, counters(key)) }
```

**Comment by dkristian** [comment]
👍 (I assume you'll be squashing commits)

The reason for the broken merge makes me a bit sad.

**Pull request merged**